src/utilities/general.c

```diff
@@ -73,6 +73,9 @@ hypre_HandleDestroy(hypre_Handle *hypre_handle_)
    hypre_HandleDeviceData(hypre_handle_) = NULL;
 #endif
 
+   /* Free error handler */
```

Comment:
It makes sense to do this here as well since users may ignore any generated error messages.  Was that the issue on Lassen, a memory leak?

Reply:
Yes, right, a memleak

Reply:
Why is it an issue on Lassen (only), not on tux instead?

Reply:
Because we don't have a test with `memtracker` on tux. We probably should add

Reply:
but we have valgrind. so it doesn't catch the leak?

Reply:
I was wondering the same thing.  We do run valgrind, but I guess we don't run a test that checks this?  It's clear that the IJ driver needs to be modified so that when `PrintErrorMessages` is called from rank 0, we also call the new `ClearErrorMessages` for the other ranks.  That should be the usual mode of operation for users, even though hypre will eventually clear everything out when the handle is destroyed.  The interface should allow for clearing-and-generating messages many times during the course of a run.

Reply:
I think the issue is that we don't run any valgrind tests that generate error messages on tux that also takes the `Print` path through the IJ driver, so there isn't any memory to clean up.  We have a `-test-error` mode for the IJ driver, but that calls `Get` from all processors so it wouldn't have a memory leak.

Reply:
Good point. Just checked manually on tux, and it actually finds the issue:

```
$ mpirun -np 2 valgrind --leak-check=yes --track-origins=yes ./ij -solver 2 -max_iter 2
==147996== Memcheck, a memory error detector
==147996== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==147996== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==147996== Command: ./ij -solver 2 -max_iter 2
==147996== 
==147997== Memcheck, a memory error detector
==147997== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==147997== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==147997== Command: ./ij -solver 2 -max_iter 2
==147997== 
=============================================
Hypre init times:
=============================================
Hypre init:
  wall clock time = 0.001402 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.001366 seconds
  cpu MFLOPS      = 0.000000


Using HYPRE_DEVELOP_STRING: v2.32.0-9-gaa1f66d63 (branch master; the develop branch)

Running with these driver parameters:
  solver ID    = 2

  Laplacian:   num_fun = 1
    (nx, ny, nz) = (10, 10, 10)
    (Px, Py, Pz) = (1, 2, 1)
    (cx, cy, cz) = (1.000000, 1.000000, 1.000000)

=============================================
Generate Matrix:
=============================================
Spatial Operator:
  wall clock time = 0.028790 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.028708 seconds
  cpu MFLOPS      = 0.000000

  Number of vector components: 1
  RHS vector has unit coefficients
  Initial guess is 0
=============================================
IJ Vector Setup:
=============================================
RHS and Initial Guess:
  wall clock time = 0.022118 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.022062 seconds
  cpu MFLOPS      = 0.000000

Solver: DS-PCG
HYPRE_ParCSRPCGGetPrecond got good precond
=============================================
Setup phase times:
=============================================
PCG Setup:
  wall clock time = 0.008703 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.008674 seconds
  cpu MFLOPS      = 0.000000

<b,b>: 1.000000e+03


Iters       ||r||_2     conv.rate  ||r||_2/||b||_2
-----    ------------   ---------  ------------ 
    1    3.651484e+01    1.154701    1.154701e+00
    2    2.779522e+01    0.761203    8.789620e-01


=============================================
Solve phase times:
=============================================
PCG Solve:
  wall clock time = 0.122544 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 0.122190 seconds
  cpu MFLOPS      = 0.000000


Iterations = 2
Final Relative Residual Norm = 8.789620e-01

0: hypre error in file "pcg.c", line 992, error code = 256 - Reached max iterations 2 in PCG before convergence
==147997== 
==147997== HEAP SUMMARY:
==147997==     in use at exit: 1,134 bytes in 1 blocks
==147997==   total heap usage: 2,940 allocs, 2,939 frees, 5,233,920 bytes allocated
==147997== 
==147996== 
==147996== HEAP SUMMARY:
==147996==     in use at exit: 0 bytes in 0 blocks
==147996==   total heap usage: 2,946 allocs, 2,946 frees, 5,234,064 bytes allocated
==147996== 
==147996== All heap blocks were freed -- no leaks are possible
==147996== 
==147996== For lists of detected and suppressed errors, rerun with: -s
==147996== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==147997== LEAK SUMMARY:
==147997==    definitely lost: 0 bytes in 0 blocks
==147997==    indirectly lost: 0 bytes in 0 blocks
==147997==      possibly lost: 0 bytes in 0 blocks
==147997==    still reachable: 1,134 bytes in 1 blocks
==147997==         suppressed: 0 bytes in 0 blocks
==147997== Reachable blocks (those to which a pointer was found) are not shown.
==147997== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==147997== 
==147997== For lists of detected and suppressed errors, rerun with: -s
==147997== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

However, if we add `--leak-check=full --show-leak-kinds=all` to the valgrind option in `runtest.sh`:

```
$ ./runtest.sh -valgrind TEST_error/error.sh
paludettomag1@tux408 ~/projects/hypre-master/src/test (master) $ ./checktest.sh 

FAILED : TEST_error/error.err  (4072)

==149144== 1,134 bytes in 1 blocks are still reachable in loss record 1 of 1
==149144==    at 0x4C39185: malloc (vg_replace_malloc.c:442)
==149144==    by 0x56D34B: hypre_HostMalloc (memory.c:266)
==149144==    by 0x56D4BC: hypre_MAlloc_core (memory.c:428)
==149144==    by 0x56DB09: hypre_MAlloc (memory.c:962)
==149144==    by 0x56DC0E: hypre_ReAlloc (memory.c:1007)
==149144==    by 0x55EDCC: hypre_error_handler (error.c:70)
==149144==    by 0x444088: hypre_PCGSolve (pcg.c:992)
==149144==    by 0x43C3A8: HYPRE_PCGSolve (HYPRE_pcg.c:47)
==149144==    by 0x4138D2: main (ij.c:5559)
==149144== 
==149167== 1,134 bytes in 1 blocks are still reachable in loss record 1 of 1
==149167==    at 0x4C39185: malloc (vg_replace_malloc.c:442)
==149167==    by 0x56D34B: hypre_HostMalloc (memory.c:266)
==149167==    by 0x56D4BC: hypre_MAlloc_core (memory.c:428)
==149167==    by 0x56DB09: hypre_MAlloc (memory.c:962)
==149167==    by 0x56DC0E: hypre_ReAlloc (memory.c:1007)
==149167==    by 0x55EDCC: hypre_error_handler (error.c:70)
==149167==    by 0x444088: hypre_PCGSolve (pcg.c:992)
==149167==    by 0x43C3A8: HYPRE_PCGSolve (HYPRE_pcg.c:47)
==149167==    by 0x4138D2: main (ij.c:5559)
==149167== 
==149370== 1,119 bytes in 1 blocks are still reachable in loss record 1 of 1
==149370==    at 0x4C39185: malloc (vg_replace_malloc.c:442)
==149370==    by 0x56D34B: hypre_HostMalloc (memory.c:266)
==149370==    by 0x56D4BC: hypre_MAlloc_core (memory.c:428)
==149370==    by 0x56DB09: hypre_MAlloc (memory.c:962)
==149370==    by 0x56DC0E: hypre_ReAlloc (memory.c:1007)
==149370==    by 0x55EDCC: hypre_error_handler (error.c:70)
==149370==    by 0x443BC1: hypre_PCGSolve (pcg.c:851)
==149370==    by 0x43C3A8: HYPRE_PCGSolve (HYPRE_pcg.c:47)
==149370==    by 0x4138D2: main (ij.c:5559)
==149370== 
==149393== 2,136 bytes in 1 blocks are still reachable in loss record 1 of 1
==149393==    at 0x4C40B77: realloc (vg_replace_malloc.c:1690)
==149393==    by 0x56DC9E: hypre_ReAlloc (memory.c:1021)
==149393==    by 0x55EDCC: hypre_error_handler (error.c:70)
==149393==    by 0x45962E: hypre_BoomerAMGSolve (par_amg_solve.c:341)
==149393==    by 0x445891: HYPRE_BoomerAMGSolve (HYPRE_parcsr_amg.c:87)
==149393==    by 0x410493: main (ij.c:4684)
==149393== 
==149190== 1,119 bytes in 1 blocks are still reachable in loss record 1 of 1
==149190==    at 0x4C39185: malloc (vg_replace_malloc.c:442)
==149190==    by 0x56D34B: hypre_HostMalloc (memory.c:266)
==149190==    by 0x56D4BC: hypre_MAlloc_core (memory.c:428)
==149190==    by 0x56DB09: hypre_MAlloc (memory.c:962)
==149190==    by 0x56DC0E: hypre_ReAlloc (memory.c:1007)
==149190==    by 0x55EDCC: hypre_error_handler (error.c:70)
==149190==    by 0x443BC1: hypre_PCGSolve (pcg.c:851)
==149190==    by 0x43C3A8: HYPRE_PCGSolve (HYPRE_pcg.c:47)
==149190==    by 0x4138D2: main (ij.c:5559)
==149190== 
==149213== 1,119 bytes in 1 blocks are still reachable in loss record 1 of 1
==149213==    at 0x4C39185: malloc (vg_replace_malloc.c:442)
==149213==    by 0x56D34B: hypre_HostMalloc (memory.c:266)
==149213==    by 0x56D4BC: hypre_MAlloc_core (memory.c:428)
==149213==    by 0x56DB09: hypre_MAlloc (memory.c:962)
==149213==    by 0x56DC0E: hypre_ReAlloc (memory.c:1007)
==149213==    by 0x55EDCC: hypre_error_handler (error.c:70)
==149213==    by 0x443BC1: hypre_PCGSolve (pcg.c:851)
==149213==    by 0x43C3A8: HYPRE_PCGSolve (HYPRE_pcg.c:47)
==149213==    by 0x4138D2: main (ij.c:5559)
==149213== 
==149236== 1,091 bytes in 1 blocks are still reachable in loss record 1 of 1
==149236==    at 0x4C39185: malloc (vg_replace_malloc.c:442)
==149236==    by 0x56D34B: hypre_HostMalloc (memory.c:266)
==149236==    by 0x56D4BC: hypre_MAlloc_core (memory.c:428)
==149236==    by 0x56DB09: hypre_MAlloc (memory.c:962)
==149236==    by 0x56DC0E: hypre_ReAlloc (memory.c:1007)
==149236==    by 0x55EDCC: hypre_error_handler (error.c:70)
==149236==    by 0x45962E: hypre_BoomerAMGSolve (par_amg_solve.c:341)
==149236==    by 0x445891: HYPRE_BoomerAMGSolve (HYPRE_parcsr_amg.c:87)
==149236==    by 0x410493: main (ij.c:4684)
==149236== 
```

Reply:
We need to tread lightly here before adding new options to valgrind in `runtest.sh`. Sometimes we do things for a reason, but maybe they should be added.  I would definitely test everything with the new options just to be sure before permanently adding them.

Reply:
> I would definitely test everything with the new options just to be sure before permanently adding them.

Agreed, let me test before adding anything

Reply:
I tested with [816c9d9](https://github.com/hypre-space/hypre/pull/1190/commits/816c9d920e1205ce11bf9ded7b8640335b76da22) and no surprises

After this change, I found a couple of other leaks in the drivers and fixed them